File: spoj/346-COINS/solution_346.cpp
#include <iostream>
#include <map>
#include <cstdio>

using namespace std;

long long coinExchange(long long num);

map<int, long long> memory;

int main()
{
    int num;
    while (cin >> num)
        cout << coinExchange(num) << "\n";
    return 0;
}

long long coinExchange(long long num)
{
    if (num == 0)
        return 0;

    if (memory.count(num) == 0)
    {
        int n1 = num / 2;
        int n2 = num / 3;
        int n3 = num / 4;

        int total = n1 + n2 + n3;

        if (total < num)
            return num;

        long long val = coinExchange(n1) + coinExchange(n2) + coinExchange(n3);
        memory[num] = val;
    }

    return memory[num];
}

File: spoj/346-COINS/solution_346.py
memory = dict()


def coin_ex(num):
    if num == 0:
        return 0

    if not memory.get(num):
        total = num // 2 + num // 3 + num // 4
        if num > total:
            return num

        val = coin_ex(num // 2) + coin_ex(num // 3) + coin_ex(num // 4)

        memory[num] = val

    return memory[num]


if __name__ == "__main__":
    while True:
        try:
            num = int(input())
            print(coin_ex(num))
        except:
            break
